Coordinated tombstone removal

bashtanov opened the pull request:
On each replica in the raft group remember and periodically update maximal cleanly compacted log offset. Calculate maximal tombstone removal offset as a minimum of these across all replicas in the configuration. Distribute it back to the nodes and restrict tombstone removal to only happen below this offset.

Offset exchange is performed via new separate RPC calls, as opposed to piggybacking on append_entries as
1) append_entries is in hot path and should not be loaded with unrelated data and logic;
2) in an otherwise idle group additional motions would be needed to trigger one, especially from the follower side.

This will ensure tombstones are never removed before corresponding preceding data records, including situations where compacted log from one replica is sent to another replica during recovery.

See [RFC here](https://docs.google.com/document/d/1xonretiC9XjoLmudixVIe_PdRalil4SJSNhG_Z0lHyM/edit?tab=t.0#heading=h.ojw25rni4gzp).

<!--
See https://github.com/redpanda-data/redpanda/blob/dev/CONTRIBUTING.md#pull-request-body
for more details and examples of what is expected in a PR body.

Content in this top section is REQUIRED. Describe, in plain language, the motivation
behind the change (bug fix, feature, improvement) in this PR and how the included
commits address it.

Add the GitHub keyword `Fixes` to link to bug(s) this PR will fix, e.g.
  Fixes #ISSUE-NUMBER, Fixes #ISSUE-NUMBER, ...

If this PR is a backport, link to the original with `Backport of PR`, e.g.
  Backport of PR #PR-NUMBER
-->

## Backports Required

<!-- Checking at least one of the checkboxes is REQUIRED if this PR is not a backport. -->

- [x] none - not a bug fix
- [ ] none - this is a backport
- [ ] none - issue does not exist in previous branches
- [ ] none - papercut/not impactful enough to backport
- [ ] v25.2.x
- [ ] v25.1.x
- [ ] v24.3.x

## Release Notes
### Bug Fixes
* Eliminated a corner case where compaction, tombstones removal and replica recovery combined would result in a wrongly present data message instead of a tombstone.

<!--
If the changes in this PR do not need to be mentioned in the release
notes, then don't add a sub-section and simply list `none`, e.g.

* none

Otherwise, adding a sub-section or `none` is REQUIRED if the PR is not a backport PR.
If this is a backport PR, adding contents to this section will override
the release notes section inherited from the original PR to dev.

Add one or more of the sub-sections with a short description bullet
point of the change, e.g.

### Bug Fixes

* Short description of the bug fix if this is a PR to `dev` branch.

### Features

* Short description of the feature. Explain how to configure.

### Improvements

* Short description of how this PR improves existing behavior.

-->

bashtanov committed:
raft: move follower_index_metadata from types.h to follower_stats.h

For coordinated compaction we need to add fields
to follower_index_metadata. The type of one of these fields depends on
the class to be added, which depends on an RPC type defined in types.h.

This would introduce a circular header dependency
if follower_index_metadata remained in types.h.
Move it to follower_stats.h

bashtanov committed:
storage/stm_manager: pass max tombstone removal offset from raft

Compaction config now contains max tombstone removal offset to config
(see #28027). Add a field for this value to storage/stm_manager in order
to pass it from raft to storage.

bashtanov committed:
ssx/abort_source: reliably subscribe to an abort_source

ss::abort_source::subscribe does not run the function if the
abort_source has already been triggered. Add a wrapper that does.

bashtanov committed:
features/table: add coordinated_compaction flag

This is for tombstone removals hoirizon on nodes to depend on compaction
state on other nodes. Flag is needed because of new raft RPC commands
introduced.

bashtanov committed:
raft/consensus: fix comment typo

bashtanov committed:
raft/tests: coordinated compaction

bashtanov committed:
raft/rpc: add RPC types and boilerplate for coordinated compaction

For coordinated tombstone removals we need raft replicas to exchange
information about their logs compaction progress. Create RPC data types
and utility functions for them.

bashtanov committed:
raft/follower_stats: add per-follower fields for coordinated compaction

For coordinated compaction we store the following info for each
follower:
- its maximum cleanly compacted offset
- RPC sender states (designed to abort relevant in-flight operations on
destruction, see ssx/single_fiber_executor for details)

bashtanov commented on src/v/ssx/abort_source.h:
```diff
@@ -133,4 +133,15 @@ class composite_abort_source {
+auto subscribe(ss::abort_source& as, Func&& func) {
+    auto subscription = as.subscribe(std::forward<Func>(func));
+    if (!subscription) {
+        subscription->on_abort(as.abort_requested_exception_ptr());
```
no, `abort_source::subscribe` allows running `on_abort` on an unsuccessful subscription, this is documented in Seastar

bashtanov commented on src/v/raft/compaction_coordinator_rpc_sender.cc:
```diff
@@ -0,0 +1,130 @@
+                      ss::future<result<report_compaction_mcco_reply>>&& fut) {
+          return process_reply(
+            std::move(fut),
+            fmt::format("MCCO={} from leader={}", mcco, leader_id));
```
nope, this is about the reply, it comes FROM there

bashtanov commented on src/v/raft/compaction_coordinator_rpc_sender.cc:
```diff
@@ -0,0 +1,130 @@
+          ss::future<result<distribute_compaction_mtro_reply>>&& fut) {
+            return process_reply(
+              std::move(fut),
+              fmt::format("MTRO={} from follower={}", mtro, follower_id));
```
nope, this is about the reply, it comes FROM there

bashtanov committed:
raft/follower_stats: fix typos in comments

dotnwat commented on src/v/ssx/abort_source.h:
```diff
@@ -133,4 +133,15 @@ class composite_abort_source {
+/// ss::abort_source::subscribe this function will invoke the callback if the
+/// abort source is already aborted when subscribed.
+template<typename Func>
+auto subscribe(ss::abort_source& as, Func&& func) {
```
should this have a different name like `subscribe_or_abort` since `abort_source::subscribe` already has the other behavior?

WillemKauf commented on src/v/raft/consensus_client_protocol.h:
```diff
@@ -131,6 +143,23 @@ class consensus_client_protocol final {
+    }
+
+    ss::future<result<distribute_compaction_mtro_reply>>
+    distribute_compaction_mtro(
```
function comments needed to describe use and purpose (imagine anyone else is looking over this file with no context around compaction, why it needs to be coordinated, what in the world an "mtro" is, etc).

bashtanov replied:
This is one of the 17 places in 10 files where we need to write absolutely identical code for each raft RPC message type. Some of them may be eliminated by templating or a bit of redesign, but 5-10 of them will inevitably have to stay. We need to agree which one(s) to comment.

> imagine anyone else is looking over this file with no context around compaction, why it needs to be coordinated, what in the world an "mtro" is, etc

I normally resolve it by searching for usages of RPC request and/or reply types. I can then see their definitions which may be commented, as well as the points where request/reply objects are created and consumed. These particular RPC calls originate in `compaction_coordinator_rpc_sender` (which refers to `compaction_coordinator` for the overview), and are processed in `consensus` where @mmaslankaprv already asked for a more detailed comment which I will provide. Do you find it reasonable?

WillemKauf replied:
> I normally resolve it by 
> [1] searching for usages of RPC request and/or reply types. I can then 
> [2] see their definitions which may be commented, as well as
> [3] the points where request/reply objects are created and consumed. 

That seems to me like 3 steps to grokking a code path instead of 

> [1] reading a code comment that is immediately available.

Happy with wherever you and Michal think is best to describe all these concepts, but at the very least I'd suggest adding code comments in other code sites like this pointing the reader to where exactly this descriptive comment resides :+1:

bashtanov replied:
FYI In the PR description I added a justification on why we added new RPC calls instead of piggy-backing on existing ones.

> That seems to me like 3 steps to grokking a code path instead of [1] reading a code comment that is immediately available.

RPC layer, especially together with Raft-specific wrappers, is too thick to traverse like this. It contains a lot of generated and boilerplate code. It operates abstractions: a request of the request type is sent to some endpoint, and a reply of the reply type is returned. Although these files, as well as those auto-generated from JSONs, contain individual functions for each RPC call name, they work with all RPC endpoints in a generic way. I'm personally not a fan of such code style and in my view it should be more generated or templated with less PImpl's, but that's what we have at the moment.

Each function does here does the same: sends a request and returns a reply. If a reader is interested in the subject matter of a particular RPC call they are probably already coming from where the function is called. But in my view a (rare) reader of this file is rather exploring raft-RPC machinery in general.

Absence of individual function comments here actually helps me, as a reader, realize I'm in the RPC machinery land already: I can see identical code in the functions and no specific logic. That said, I don't oppose adding much more verbose comments on the _classes_ in this file as well as in `raft/service.h`, `raft/rpc_client_protocol.h`, `raft/buffered_protocol.h`.

@mmaslankaprv WDYT?

mmaslankaprv replied:
I think comments in this file, defining the protocol makes sens. I am sorry i didn't add them earlier. Maybe we can do this in a follow up PR. We will add comments to all the protocol methods. This way the protocol definition file will become a source of truth for the Raft RPC protocol documentation definition.

bashtanov replied:
Thanks, I'll create a follow-up PR to describe all raft RPC endpoints in one file.

WillemKauf commented on src/v/raft/compaction_coordinator_rpc_sender.cc:
```diff
@@ -0,0 +1,130 @@
+        return ssx::now(ss::stop_iteration::yes);
+    }
+    vlog(_logger.trace, "sending MCCO={} to leader={}", mcco, leader_id);
+    return _client_protocol
```
nit: nicer as a coroutine?

bashtanov replied:
I tried back and forth between `ss::coroutine::as_future` and `then_wrapped`. It looks a bit nicer with `ss::coroutine::as_future`, but it still does not make sense to make this function a coroutine. I'll change it.

WillemKauf commented on src/v/storage/disk_log_impl.cc:
```diff
@@ -4604,7 +4604,9 @@ disk_log_impl::earliest_removable_timestamp(model::offset o) const {
           = config().tombstone_retention_ms().has_value()
             && seg->has_clean_compact_timestamp()
-            && seg->index().may_have_tombstone_records();
+            && seg->index().may_have_tombstone_records()
```
This function here is actually dead code now. I'd rather you remove it.

See `can_discard()` in `segment_utils.h` for the location of where we actually need to plumb this new `max_tombstone_remove_offset()`

bashtanov replied:
> This function here is actually dead code now. I'd rather you remove it.

will do

> See can_discard() in segment_utils.h for the location of where we actually need to plumb this new max_tombstone_remove_offset()

It's already there via `storage::internal::is_past_tombstone_delete_horizon()`

WillemKauf replied:
Oh, I forgot this (`It's already there via storage::internal::is_past_tombstone_delete_horizon()`) was a part of the previous PR, thanks :+1:

WillemKauf reviewed (commented):
only making it halfway through tonight. Some small comments (going to tackle the meat tomorrow).

Seems like a lot of new infrastructure in the `raft` layer to support this out-of-band communication of two offsets, which feels unfortunate.

mmaslankaprv commented on src/v/raft/consensus.h:
```diff
@@ -572,6 +578,14 @@ class consensus {
+    ss::future<report_compaction_mcco_reply>
+      process_compaction_mcco(report_compaction_mcco_request);
+
+    // Function invoked on follower side to receive MTRO from leader.
```
Please add comments what mtro and mcco means

mmaslankaprv commented on src/v/raft/compaction_coordinator_rpc_sender.cc:
```diff
@@ -0,0 +1,130 @@
+    if (_self != _leader_id) {
+        vlog(
+          _logger.trace,
+          "leader is now {}, stop sending MTRO={} to follower={}",
```
nit: in whole `raft` submodule we use `:` instead of `=` to separate variable from its name

mmaslankaprv commented on src/v/raft/follower_stats.h:
```diff
@@ -13,10 +13,142 @@
+      = decltype(std::declval<compaction_coordinator_rpc_sender>()
+                   .make_mtro_sender(
+                     std::declval<vnode>(), std::declval<model::offset>()));
+    std::unique_ptr<ssx::single_fiber_executor<mtro_sender_t>> mtro_sender;
```
can we store this sender in a compaction coordinator ?

bashtanov replied:
These per-follower senders need to come and go as configuration changes. We can store them in compaction coordinator, but this will require compaction coordinator to have one more map of per-follower states, and keep it in sync with the one stored in follower stats. IIRC we discussed it once agreed to keep it in follower stats if I rename it. I'll rename it in a follow-up PR.

mmaslankaprv commented on src/v/raft/recovery_stm.h:
```diff
@@ -12,6 +12,7 @@
 #pragma once
 
 #include "base/outcome.h"
+#include "raft/follower_stats.h"
```
why do we need this include ?

bashtanov replied:
because `recovery_stm.h` uses `follower_index_metadata`, which is now defined in `raft/follower_stats.h`, not in `raft/types.h`

bashtanov committed:
storage/log: remove log::earliest_removable_timestamp() (dead code)

This function is not in use, and will not be needed after coordinated
tombstone removal is enabled.

bashtanov commented on src/v/ssx/abort_source.h:
```diff
@@ -133,4 +133,15 @@ class composite_abort_source {
+/// immediately if the abort source is already aborted.
+template<typename Func>
+auto subscribe_or_trigger(ss::abort_source& as, Func&& func) {
+    auto subscription = as.subscribe(std::forward<Func>(func));
+    if (!subscription) {
+        subscription->on_abort(as.abort_requested_exception_ptr());
+    }
+    return subscription;
```
I'll resolve your concerns with a unit test

vbotbuildovich commented:
<!--TEST RESULTS-->
<!--4929adf7-700b-40c8-b024-358754d2a569-->
<!--TEST RESULTS-->
<!--29be7fc8-3f07-4562-925a-cf0b1ebfe3a7-->
<!--TEST RESULTS-->
<!--19a86076-579f-4ee4-8d95-8962207d6bd9-->
<!--TEST RESULTS-->
<!--fe7a2b12-86e5-44a3-94df-4c93b322bec7-->
<!--TEST RESULTS-->
<!--9c1f9094-f913-47f9-82dd-04f1b80e4ece-->
<!--title-->
## CI test results
<!--/title-->
<!--body-->
<details>
<summary>test results on <a href="https://buildkite.com/redpanda/redpanda/builds/74422">build#74422</a></summary>


| test_class                       | test_method                    | test_arguments                                                                    | test_kind   | job_url                                                                                   | test_status   | passed   | reason                                                                                                                                                                   | test_history                                                                                                                                                               |
|:---------------------------------|:-------------------------------|:----------------------------------------------------------------------------------|:------------|:------------------------------------------------------------------------------------------|:--------------|:---------|:-------------------------------------------------------------------------------------------------------------------------------------------------------------------------|:---------------------------------------------------------------------------------------------------------------------------------------------------------------------------|
| ShadowLinkingReplicationTests    | test_replication_basic         | {"shuffle_leadership": true, "source_cluster_spec": {"cluster_type": "redpanda"}} | integration | https://buildkite.com/redpanda/redpanda/builds/74422#0199f32d-4d56-4e32-9dec-1ea8a2130211 | FLAKY         | 14/21    | upstream reliability is '96.59574468085106'. current run reliability is '66.66666666666666'. drift is 29.92908 and the allowed drift is set to 50. The test should PASS  | https://redpanda.metabaseapp.com/dashboard/87-tests?tab=142-dt-individual-test-history&test_class=ShadowLinkingReplicationTests&test_method=test_replication_basic         |
| ShadowLinkingReplicationTests    | test_replication_basic         | {"shuffle_leadership": true, "source_cluster_spec": {"cluster_type": "redpanda"}} | integration | https://buildkite.com/redpanda/redpanda/builds/74422#0199f337-37d9-4285-926f-e361ab405f8d | FLAKY         | 19/21    | upstream reliability is '96.58848614072495'. current run reliability is '90.47619047619048'. drift is 6.1123 and the allowed drift is set to 50. The test should PASS    | https://redpanda.metabaseapp.com/dashboard/87-tests?tab=142-dt-individual-test-history&test_class=ShadowLinkingReplicationTests&test_method=test_replication_basic         |
| ShadowLinkingReplicationTests    | test_replication_with_failures | null                                                                              | integration | https://buildkite.com/redpanda/redpanda/builds/74422#0199f337-37da-48b6-a0e3-5c1a7d4304ec | FLAKY         | 8/21     | upstream reliability is '81.73652694610777'. current run reliability is '38.095238095238095'. drift is 43.64129 and the allowed drift is set to 50. The test should PASS | https://redpanda.metabaseapp.com/dashboard/87-tests?tab=142-dt-individual-test-history&test_class=ShadowLinkingReplicationTests&test_method=test_replication_with_failures |
| WriteCachingFailureInjectionTest | test_unavoidable_data_loss     | null                                                                              | integration | https://buildkite.com/redpanda/redpanda/builds/74422#0199f337-37d7-4ffe-a542-c7e575c8ccf7 | FLAKY         | 19/21    | upstream reliability is '90.80459770114942'. current run reliability is '90.47619047619048'. drift is 0.32841 and the allowed drift is set to 50. The test should PASS   | https://redpanda.metabaseapp.com/dashboard/87-tests?tab=142-dt-individual-test-history&test_class=WriteCachingFailureInjectionTest&test_method=test_unavoidable_data_loss  |
</details>
<details>
<summary>test results on <a href="https://buildkite.com/redpanda/redpanda/builds/74541">build#74541</a></summary>


| test_class                    | test_method                                             | test_arguments                               | test_kind   | job_url                                                                                   | test_status   | passed   | reason                                                                                                                                                                 | test_history                                                                                                                                                                                   |
|:------------------------------|:--------------------------------------------------------|:---------------------------------------------|:------------|:------------------------------------------------------------------------------------------|:--------------|:---------|:-----------------------------------------------------------------------------------------------------------------------------------------------------------------------|:-----------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------|
| ShadowLinkBasicTests          | test_rapid_shadow_link_toggling                         | null                                         | integration | https://buildkite.com/redpanda/redpanda/builds/74541#019a02e0-d7a1-4439-8483-4c737a3e5d5d | FLAKY         | 19/21    | upstream reliability is '98.4447900466563'. current run reliability is '90.47619047619048'. drift is 7.9686 and the allowed drift is set to 50. The test should PASS   | https://redpanda.metabaseapp.com/dashboard/87-tests?tab=142-dt-individual-test-history&test_class=ShadowLinkBasicTests&test_method=test_rapid_shadow_link_toggling                             |
| ShadowLinkingReplicationTests | test_replication_with_failures                          | null                                         | integration | https://buildkite.com/redpanda/redpanda/builds/74541#019a02ee-9d93-4e2b-baaf-be6a06946286 | FLAKY         | 12/21    | upstream reliability is '63.13173156586578'. current run reliability is '57.14285714285714'. drift is 5.98887 and the allowed drift is set to 50. The test should PASS | https://redpanda.metabaseapp.com/dashboard/87-tests?tab=142-dt-individual-test-history&test_class=ShadowLinkingReplicationTests&test_method=test_replication_with_failures                     |
| DataMigrationsApiTest         | test_higher_level_migration_api                         | null                                         | integration | https://buildkite.com/redpanda/redpanda/builds/74541#019a02e0-d7a1-4439-8483-4c737a3e5d5d | FLAKY         | 20/21    | upstream reliability is '99.24385633270322'. current run reliability is '95.23809523809523'. drift is 4.00576 and the allowed drift is set to 50. The test should PASS | https://redpanda.metabaseapp.com/dashboard/87-tests?tab=142-dt-individual-test-history&test_class=DataMigrationsApiTest&test_method=test_higher_level_migration_api                            |
| NodesDecommissioningTest      | test_decommissioning_finishes_after_manual_cancellation | {"cloud_topic": false, "delete_topic": true} | integration | https://buildkite.com/redpanda/redpanda/builds/74541#019a02ee-9d91-4297-be7a-b26a73859b82 | FLAKY         | 19/21    | upstream reliability is '97.79735682819384'. current run reliability is '90.47619047619048'. drift is 7.32117 and the allowed drift is set to 50. The test should PASS | https://redpanda.metabaseapp.com/dashboard/87-tests?tab=142-dt-individual-test-history&test_class=NodesDecommissioningTest&test_method=test_decommissioning_finishes_after_manual_cancellation |
</details>
<details>
<summary>test results on <a href="https://buildkite.com/redpanda/redpanda/builds/74655">build#74655</a></summary>


| test_class                          | test_method           | test_arguments              | test_kind   | job_url                                                                                   | test_status   | passed   | reason                                                                                                                                                                   | test_history                                                                                                                                                     |
|:------------------------------------|:----------------------|:----------------------------|:------------|:------------------------------------------------------------------------------------------|:--------------|:---------|:-------------------------------------------------------------------------------------------------------------------------------------------------------------------------|:-----------------------------------------------------------------------------------------------------------------------------------------------------------------|
| DataMigrationsApiTest               | test_mount_inexistent | null                        | integration | https://buildkite.com/redpanda/redpanda/builds/74655#019a0975-0cd5-47a9-89bb-4457b2304f7b | FLAKY         | 20/21    | upstream reliability is '99.78632478632478'. current run reliability is '95.23809523809523'. drift is 4.54823 and the allowed drift is set to 50. The test should PASS   | https://redpanda.metabaseapp.com/dashboard/87-tests?tab=142-dt-individual-test-history&test_class=DataMigrationsApiTest&test_method=test_mount_inexistent        |
| ShadowLinkingRandomOpsTest          | test_node_operations  | {"failures": false}         | integration | https://buildkite.com/redpanda/redpanda/builds/74655#019a0975-0cd6-45a2-a368-2e63b84d98b2 | FLAKY         | 10/21    | upstream reliability is '86.95652173913044'. current run reliability is '47.61904761904761'. drift is 39.33747 and the allowed drift is set to 50. The test should PASS  | https://redpanda.metabaseapp.com/dashboard/87-tests?tab=142-dt-individual-test-history&test_class=ShadowLinkingRandomOpsTest&test_method=test_node_operations    |
| ShadowLinkingRandomOpsTest          | test_node_operations  | {"failures": false}         | integration | https://buildkite.com/redpanda/redpanda/builds/74655#019a0980-03cc-40fa-9163-2edee2be34bd | FLAKY         | 6/21     | upstream reliability is '40.4040404040404'. current run reliability is '28.57142857142857'. drift is 11.83261 and the allowed drift is set to 50. The test should PASS   | https://redpanda.metabaseapp.com/dashboard/87-tests?tab=142-dt-individual-test-history&test_class=ShadowLinkingRandomOpsTest&test_method=test_node_operations    |
| ShadowLinkingRandomOpsTest          | test_node_operations  | {"failures": true}          | integration | https://buildkite.com/redpanda/redpanda/builds/74655#019a0975-0ccb-4319-a71d-70715ead5d64 | FAIL          | 0/1      |                                                                                                                                                                          | https://redpanda.metabaseapp.com/dashboard/87-tests?tab=142-dt-individual-test-history&test_class=ShadowLinkingRandomOpsTest&test_method=test_node_operations    |
| ShadowLinkingRandomOpsTest          | test_node_operations  | {"failures": true}          | integration | https://buildkite.com/redpanda/redpanda/builds/74655#019a0980-03b8-4d1a-befb-b1134be8c3d4 | FLAKY         | 4/21     | upstream reliability is '33.33333333333333'. current run reliability is '19.047619047619047'. drift is 14.28571 and the allowed drift is set to 50. The test should PASS | https://redpanda.metabaseapp.com/dashboard/87-tests?tab=142-dt-individual-test-history&test_class=ShadowLinkingRandomOpsTest&test_method=test_node_operations    |
| WriteCachingFailureInjectionE2ETest | test_crash_all        | {"use_transactions": false} | integration | https://buildkite.com/redpanda/redpanda/builds/74655#019a0980-03c8-4d3a-be43-0efbe6ac1b43 | FLAKY         | 16/21    | upstream reliability is '81.72043010752688'. current run reliability is '76.19047619047619'. drift is 5.52995 and the allowed drift is set to 50. The test should PASS   | https://redpanda.metabaseapp.com/dashboard/87-tests?tab=142-dt-individual-test-history&test_class=WriteCachingFailureInjectionE2ETest&test_method=test_crash_all |
</details>
<details>
<summary>test results on <a href="https://buildkite.com/redpanda/redpanda/builds/74666">build#74666</a></summary>


| test_class           | test_method                     | test_arguments   | test_kind   | job_url                                                                                   | test_status   | passed   | reason                                                                                                                                                                 | test_history                                                                                                                                                       |
|:---------------------|:--------------------------------|:-----------------|:------------|:------------------------------------------------------------------------------------------|:--------------|:---------|:-----------------------------------------------------------------------------------------------------------------------------------------------------------------------|:-------------------------------------------------------------------------------------------------------------------------------------------------------------------|
| ShadowLinkBasicTests | test_rapid_shadow_link_toggling | null             | integration | https://buildkite.com/redpanda/redpanda/builds/74666#019a0a9b-f12b-4d1a-8bf7-3896e5678ade | FLAKY         | 14/21    | upstream reliability is '94.65116279069767'. current run reliability is '66.66666666666666'. drift is 27.9845 and the allowed drift is set to 50. The test should PASS | https://redpanda.metabaseapp.com/dashboard/87-tests?tab=142-dt-individual-test-history&test_class=ShadowLinkBasicTests&test_method=test_rapid_shadow_link_toggling |
</details>
<details>
<summary>test results on <a href="https://buildkite.com/redpanda/redpanda/builds/74684">build#74684</a></summary>


| test_class                 | test_method          | test_arguments        | test_kind   | job_url                                                                                   | test_status   | passed   | reason                                                                                                                                                                  | test_history                                                                                                                                                  |
|:---------------------------|:---------------------|:----------------------|:------------|:------------------------------------------------------------------------------------------|:--------------|:---------|:------------------------------------------------------------------------------------------------------------------------------------------------------------------------|:--------------------------------------------------------------------------------------------------------------------------------------------------------------|
| PartitionMovementTest      | test_empty           | {"num_to_upgrade": 0} | integration | https://buildkite.com/redpanda/redpanda/builds/74684#019a0c03-8da9-48f9-8b0f-901a5ffcbee0 | FLAKY         | 17/21    | upstream reliability is '100.0'. current run reliability is '80.95238095238095'. drift is 19.04762 and the allowed drift is set to 50. The test should PASS             | https://redpanda.metabaseapp.com/dashboard/87-tests?tab=142-dt-individual-test-history&test_class=PartitionMovementTest&test_method=test_empty                |
| PartitionMovementTest      | test_static          | {"num_to_upgrade": 0} | integration | https://buildkite.com/redpanda/redpanda/builds/74684#019a0c09-0276-47d3-9e3e-01a39d8400e6 | FLAKY         | 18/21    | upstream reliability is '100.0'. current run reliability is '85.71428571428571'. drift is 14.28571 and the allowed drift is set to 50. The test should PASS             | https://redpanda.metabaseapp.com/dashboard/87-tests?tab=142-dt-individual-test-history&test_class=PartitionMovementTest&test_method=test_static               |
| ShadowLinkingRandomOpsTest | test_node_operations | {"failures": false}   | integration | https://buildkite.com/redpanda/redpanda/builds/74684#019a0c09-0278-40ef-903b-2c52acccf54b | FLAKY         | 7/21     | upstream reliability is '35.66433566433567'. current run reliability is '33.33333333333333'. drift is 2.331 and the allowed drift is set to 50. The test should PASS    | https://redpanda.metabaseapp.com/dashboard/87-tests?tab=142-dt-individual-test-history&test_class=ShadowLinkingRandomOpsTest&test_method=test_node_operations |
| ShadowLinkingRandomOpsTest | test_node_operations | {"failures": true}    | integration | https://buildkite.com/redpanda/redpanda/builds/74684#019a0c03-8da4-4aa6-9d9a-2984013e8bbf | FAIL          | 0/1      |                                                                                                                                                                         | https://redpanda.metabaseapp.com/dashboard/87-tests?tab=142-dt-individual-test-history&test_class=ShadowLinkingRandomOpsTest&test_method=test_node_operations |
| ShadowLinkingRandomOpsTest | test_node_operations | {"failures": true}    | integration | https://buildkite.com/redpanda/redpanda/builds/74684#019a0c09-026d-49bd-b926-44f6bf3c711c | FLAKY         | 1/21     | upstream reliability is '33.33333333333333'. current run reliability is '4.761904761904762'. drift is 28.57143 and the allowed drift is set to 50. The test should PASS | https://redpanda.metabaseapp.com/dashboard/87-tests?tab=142-dt-individual-test-history&test_class=ShadowLinkingRandomOpsTest&test_method=test_node_operations |
</details>
<!--/body-->

joe-redpanda commented on src/v/raft/buffered_protocol.h:
```diff
@@ -155,6 +155,15 @@ class buffered_protocol : public consensus_client_protocol::impl {
     ss::future<result<remake_learner_state_reply>> remake_learner_state(
       model::node_id, remake_learner_state_request, rpc::client_opts) final;
 
+    ss::future<result<report_compaction_mcco_reply>> report_compaction_mcco(
```
Can you add acronym definitions

bashtanov replied:
See [here](https://github.com/redpanda-data/redpanda/pull/28087/files#diff-040e4eb5009543030cfd90a36e8a880b7eb2a643fcf7e2d249ba44e245279884R30-R32) for acronym definitions.
See [this discussion](https://github.com/redpanda-data/redpanda/pull/28087#discussion_r2438126250) for why I think adding comments about it to raft-rpc machinery is not a good idea.

joe-redpanda commented on src/v/raft/tests/coordinated_compaction_test.cc:
```diff
@@ -0,0 +1,207 @@
+        return recovered_offset >= last_data_offset;
+    });
+
+    co_await ss::sleep(1s);
```
What is the purpose of this sleep?

bashtanov replied:
I added comments to all sleeps in the test.

bashtanov commented:
@WillemKauf @bharathv @joe-redpanda I would appreciate if you could have a look

WillemKauf commented on src/v/raft/compaction_coordinator.h:
```diff
@@ -0,0 +1,142 @@
+    // RPC sender for mcco updates to the leader
+    using mcco_sender_t = decltype(_rpc_sender.make_local_mcco_sender(
+      std::declval<vnode>(), std::declval<model::offset>()));
+    ssx::single_fiber_executor<mcco_sender_t> _leader_sender;
```
general Q: why does this require a single fiber executor instead of a work queue?

bashtanov replied:
These operations/requests are idempotent within each executor as MCCO and MTRO never decrease. Thus, completing an earlier request is pointless (and undesired for performance reasons) if a new one was sent.

WillemKauf commented on src/v/raft/tests/coordinated_compaction_test.cc:
```diff
@@ -0,0 +1,350 @@
+    });
+}
+
+TEST_F_CORO(coco_fixture, test_leadership_change) {
```
say we had an incident in which tombstone removal appeared to be blocked- is there an easy way to report MCCO per broker? Might be worth adding a metric or admin endpoint for? 

(may also make for a good e2e ducktape test, though this is not a requirement for this PR per se)

bashtanov replied:
I'll expose MCCO & MTRO in a debug endpoint. I'm also looking to add a ducktape test. Both in follow up PRs due to time pressure.

WillemKauf reviewed (dismissed):
From the storage perspective this LGTM, a lot of Raft infrastructure that seems okay to me but please defer to replication team weigh-in on those aspects

joe-redpanda commented on src/v/raft/compaction_coordinator.cc:
```diff
@@ -0,0 +1,296 @@
+        return;
+    }
+    update_mtro(new_mtro);
+    // we may be a follower lagging behind, so our MCCO or even max offset may
```
learner?

I don't think non-learner followers can lag behind mtro

bashtanov replied:
yes, thanks

joe-redpanda reviewed (dismissed):
Nice

bharathv commented on src/v/raft/compaction_coordinator.h:
```diff
@@ -0,0 +1,142 @@
+      std::declval<vnode>(), std::declval<model::offset>()));
+    ssx::single_fiber_executor<mcco_sender_t> _leader_sender;
+    // model::offset{} means never calculated
+    model::offset _local_mcco;
+    // model::offset{} if never successfully calculated, pending remote MCCOs
+    model::offset _mtro;
+
```
can we wire these into the partition debug end point? ( `v1/debug/partitions/kafka/foo/<id>`), makes debugging much easier if retention don't make progress (fine to make it in a separate PR).

mmaslankaprv replied:
this is great suggestion

bharathv commented on src/v/raft/compaction_coordinator.cc:
```diff
@@ -0,0 +1,296 @@
+
+void compaction_coordinator::recalculate_mtro() {
+    model::offset new_mtro = _local_mcco;
+    for (const auto& [node_id, fstat] : _fstats) {
```
just to make sure I get it, so the mtro only moves if _all_ the followers (voters and learners) advanced `mcco` right.

bashtanov replied:
yes, and the leader as well

bharathv commented on src/v/raft/consensus.cc:
```diff
@@ -4428,4 +4441,43 @@ consensus::do_snapshot_and_truncate_log(model::offset truncation_point) {
+    _compaction_coordinator.update_mtro_from_leader(req.mtro);
+    co_return distribute_compaction_mtro_reply{
+      .success = distribute_compaction_mtro_reply::is_success::yes};
+}
```
would be nice if these can be encapsulated within compaction_coordinator. I'm a bit worried that the compaction is leaking into raft but I think that is unavoidable in this case, so trying to contain the spread as much as possible.

bharathv commented on src/v/raft/consensus.cc:
```diff
@@ -127,10 +127,21 @@ consensus::consensus(
   , _fstats(_self)
+  , _ctxlog(group, _log->config().ntp())
+  , _features(ft)
+  , _compaction_coordinator(
```
wondering if you thought about adding a kill switch for the whole feature (for extreme cases to revert back to old behavior, eg: not starting the coordinator  / letting storage decide its own MTRO etc).

bashtanov replied:
Not sure if we want it tbh. The whole feature is to avoid a corner case where tombstones are deleted incorrectly so that a reader may observe a data entry without corresponding tombstone. @mmaslankaprv @WillemKauf  wdyt?

bharathv reviewed (commented):
lgtm generally, a few general comments.

mmaslankaprv commented on src/v/raft/types.h:
```diff
@@ -794,6 +676,104 @@ struct remake_learner_state_reply
+    operator<<(std::ostream& o, const report_compaction_mcco_reply& r) {
+        fmt::print(o, "success: {}", r.success);
+        return o;
+    }
```
nit: use `format_to`

mmaslankaprv commented on src/v/raft/compaction_coordinator.cc:
```diff
@@ -0,0 +1,296 @@
+void compaction_coordinator::on_local_mcco_update(model::offset new_mcco) {
+    vlog(
+      _logger.debug,
+      "updating local MCCO from {} to {}",
```
i am not fan of the acronyms in log messages, can we always use full name in log lines ?

mmaslankaprv commented on src/v/raft/compaction_coordinator.cc:
```diff
@@ -0,0 +1,296 @@
+        }
+        if (fstat.max_cleanly_compacted_offset < new_mtro) {
+            new_mtro = fstat.max_cleanly_compacted_offset;
+        }
```
is that just: 
```
new_mtro = std::min(new_mtro, fstat.max_cleanly_compacted_offset)
```

mmaslankaprv commented on src/v/raft/compaction_coordinator.cc:
```diff
@@ -0,0 +1,296 @@
+          "attempted to move local MCCO backwards from {} to {}",
+          _local_mcco,
+          new_mcco);
+        return;
+    }
+    _local_mcco = new_mcco;
+
+    if (is_leader()) {
+        recalculate_mtro();
+    } else {
+        send_local_mcco_to_leader();
+    }
+}
+
+void compaction_coordinator::send_local_mcco_to_leader() {
+    if (!_leader_id) {
+        vlog(
+          _logger.trace,
+          "cannot send MCCO:{} to leader, no leader",
+          _local_mcco);
+        return;
+    }
+    vlog(
+      _logger.trace,
+      "sending local MCCO: {} to leader: {}",
+      _local_mcco,
+      _leader_id);
+    ssx::background = _leader_sender.submit(
+      _rpc_sender.make_local_mcco_sender(*_leader_id, _local_mcco));
+}
+
+void compaction_coordinator::send_mtro_to_followers() {
+    for (const auto& [node_id, fstat] : _fstats) {
+        if (node_id == _self) {
+            continue;
+        }
+        ssx::background = fstat.mtro_sender->submit(
+          _rpc_sender.make_mtro_sender(node_id, _mtro));
+    }
+}
```
Would it simpler if leader requests local MCCOs from the followers ? This way we would have communication in a single direction. Current implementation is different than rest of Raft protocol as this is the only one where the follower initiates the communication. Can leader do this periodically ? I think this way the design would be simplified

bashtanov replied:
This is going to be a serious redesign now. Let me try to find out if there are any obstacles.

bashtanov committed:
raft/compaction_coordinator: calculate safe offset to remove tombstones

Tombstone and transaction batch removals were subject to a problem where
a reader may see an live data record and then a compacted away tombstone
record with the same key. Add a component to distribute cleanly
compacted offsets between nodes, calculate max tombstone removal offset
as minimum of them, and distribute the latter back to all replicas. This
will ensure tombstones are never removed before corresponding preceding
data records, including situations where compacted log from one replica
is sent to another replica during recovery.

vbotbuildovich commented:
<!--FAILURES SLASH COMMAND https://buildkite.com/redpanda/redpanda/builds/74655-->
<!--f14327a6-7828-435e-a961-71342948e61d-->
<!--title-->
## Retry command for [Build#74655](https://buildkite.com/redpanda/redpanda/builds/74655)
please wait until all jobs are finished before running the slash command
<!--/title-->
<!--body-->
```
/ci-repeat 1
tests/rptest/tests/shadow_linking_rnot_test.py::ShadowLinkingRandomOpsTest.test_node_operations@{"failures":true}
```
<!--/body-->

bashtanov commented:
/ci-repeat 1
tests/rptest/tests/shadow_linking_rnot_test.py::ShadowLinkingRandomOpsTest.test_node_operations@{"failures":true}

vbotbuildovich commented:
<!--FAILURES SLASH COMMAND https://buildkite.com/redpanda/redpanda/builds/74684-->
<!--5ed319bd-2258-4db1-94a3-25057f10d28f-->
<!--title-->
## Retry command for [Build#74684](https://buildkite.com/redpanda/redpanda/builds/74684)
please wait until all jobs are finished before running the slash command
<!--/title-->
<!--body-->
```
/ci-repeat 1
tests/rptest/tests/shadow_linking_rnot_test.py::ShadowLinkingRandomOpsTest.test_node_operations@{"failures":true}
```
<!--/body-->

bashtanov commented:
@mmaslankaprv @bharathv @joe-redpanda @WillemKauf @dotnwat did I address your comments?
I'll make these changes in follow-up PRs:
1) rename follower stats to follower states
2) expose MCCO and MTRO in a debug endpoint
3) describe all Raft RPC endpoints in code comments in one file
4) add an end2end ducktape test

bharathv commented:
> @mmaslankaprv @bharathv @joe-redpanda @WillemKauf @dotnwat did I address your comments? I'll make these changes in follow-up PRs:
> 
> 1. rename follower stats to follower states
> 2. expose MCCO and MTRO in a debug endpoint
> 3. describe all Raft RPC endpoints in code comments in one file
> 4. add an end2end ducktape test

curious about Michal's suggestion to make it leader driven, is that still a future work too?

Other than that my comments are addressed, thanks.

mmaslankaprv commented:
> > @mmaslankaprv @bharathv @joe-redpanda @WillemKauf @dotnwat did I address your comments? I'll make these changes in follow-up PRs:
> > 
> > 1. rename follower stats to follower states
> > 2. expose MCCO and MTRO in a debug endpoint
> > 3. describe all Raft RPC endpoints in code comments in one file
> > 4. add an end2end ducktape test
> 
> curious about Michal's suggestion to make it leader driven, is that still a future work too?
> 
> Other than that my comments are addressed, thanks.

It is done already

bharathv commented:
> > > @mmaslankaprv @bharathv @joe-redpanda @WillemKauf @dotnwat did I address your comments? I'll make these changes in follow-up PRs:
> > > 
> > > 1. rename follower stats to follower states
> > > 2. expose MCCO and MTRO in a debug endpoint
> > > 3. describe all Raft RPC endpoints in code comments in one file
> > > 4. add an end2end ducktape test
> > 
> > 
> > curious about Michal's suggestion to make it leader driven, is that still a future work too?
> > Other than that my comments are addressed, thanks.
> 
> It is done already

oh ya right, I was under the impression that the suggestion was also to use a single RPC to both distribute MTRO and get MCCO (so the whole surface area is smaller).  Maybe I misunderstood the suggestion, my bad.

bashtanov commented:
> I was under the impression that the suggestion was also to use a single RPC to both distribute MTRO and get MCCO (so the whole surface area is smaller)

Currently the logic is to gather MCCOs and then to send MTROs shortly after.

If we unify them but still run the round twice it'll be a bit awkward: in the first round MTROs will be no better than in the second one, and MCCOs of the second one will only bump tombstone removal horizon on the leader.

If, however, we run it only once, the calculation+distribution won't make any progress at all if leadership changes more often that once an hour, which I would guess can easily take place even in a healthy cluster.

bharathv commented:
> > I was under the impression that the suggestion was also to use a single RPC to both distribute MTRO and get MCCO (so the whole surface area is smaller)
> 
> Currently the logic is to gather MCCOs and then to send MTROs shortly after.
> 
> If we unify them but still run the round twice it'll be a bit awkward: in the first round MTROs will be no better than in the second one, and MCCOs of the second one will only bump tombstone removal horizon on the leader.
> 
> If, however, we run it only once, the calculation+distribution won't make any progress at all if leadership changes more often that once an hour, which I would guess can easily take place even in a healthy cluster.

ack, makes sense thanks.
